cms/djangoapps/contentstore/tests/test_tasks.py

```diff
@@ -340,30 +382,179 @@ def test_course_scan_occurs_on_published_version(self, mock_modulestore, mock_mo
+        with patch("cms.djangoapps.contentstore.tasks._validate_batch", new_callable=AsyncMock) as mock_validate_batch:
+            mock_validate_batch.side_effect = lambda x, y: x
+            validated_urls = await _validate_urls_access_in_batches(url_list, course_key, batch_size)
+            mock_validate_batch.assert_called()
+            assert mock_validate_batch.call_count == 3 # two full batches and one partial batch
+            assert validated_urls == url_list, \
+                f"List of validated urls {validated_urls} is not identical to sourced urls {url_list}"
+
+    @pytest.mark.asyncio
+    async def test_all_links_are_validated_with_batch_validation(self):
+        '''
+        Here the focus is not on batching, but rather that when validation occurs it does so on the intended
+        URL strings
+        '''
+        with patch("cms.djangoapps.contentstore.tasks._validate_url_access", new_callable=AsyncMock) as mock_validate:
+            mock_validate.return_value = {"status": 200}
+
+            url_list = ['1', '2', '3', '4', '5']
+            course_key = 'course-v1:edX+DemoX+Demo_Course'
+            batch_size=2
+            await _validate_urls_access_in_batches(url_list, course_key, batch_size)
+            args_list = mock_validate.call_args_list
+            urls = [call_args.args[1] for call_args in args_list] # The middle argument in each of the function calls
+            for i in range(1,len(url_list)+1):
+                assert str(i) in urls, f'{i} not supplied as a url for validation in batches function'
 
     def test_no_retries_on_403_access_denied_links(self):
-        raise NotImplementedError
+        '''
+        No mocking required here. Will populate "filtering_input" with simulated results for link checks where
+        some links time out, some links receive 403 errors, and some receive 200 success. This test then
+        ensures that "_filter_by_status()" tallies the three categories as expected, and formats the result
+        as expected.
+        '''
+        url_list = ['1', '2', '3', '4', '5']
+        filtering_input = []
+        for i in range(1, len(url_list)+1): # Notch out one of the URLs, having it return a '403' status code
+            filtering_input.append(
+            {'block_id': f'block_{i}',
+             'url': str(i),
+             'status': 200},
+            )
+        filtering_input[2]['status'] = 403
+        filtering_input[3]['status'] = 500
+        filtering_input[4]['status'] = None
+
+        broken_or_locked_urls, retry_list = _filter_by_status(filtering_input)
+        assert len(broken_or_locked_urls) == 2  # The inputs with status = 403 and 500
+        assert len(retry_list) == 1             # The input with status = None
+        assert retry_list[0][1] == '5'      # The only URL fit for a retry operation (status == None)
+
+    # TODO - test retry logic
+
+    @pytest.mark.skip(reason="Failing but needs review -- test not yet correct")
```

Comment:
If it takes too long to get this test correct, let's remove. If this test is proven to be valuable in the future, we can add it again.